### DISCUSSION TITLE
"Calculating Liquidity" formula question

### DISCUSSION THREAD

**Discussion by Zarfsec** [Jeiwan/uniswapv3-book#38] [Milestone 1. First Swap] [answered]
in the `liquidity1()` function, amount is multiplied by q96 to convert it into a Q64.96 number. Why are we not doing this in the `liquidity0()` function (and dividing by q96 instead)?

Also, `amount_usdc` is set to `5000 * 10**18`. Since the USDC contract uses 6 decimals, shouldn't `amount_usdc` then be equal to `5000 * 10**6` instead?

**Comment by uniyj** [comment, marked as answer]
The `sqrtp`s are stored as Q64.96. To get back to a "normal" number (not in Q64.96), you need to "compensate" for that `q96` conversion.

For `liquidity0()`, you have two `sqrtp`s multiplied together in the numerator, one of them is cancelled out by the lot of `sqrtp` in the denominator, so there is still one conversion of `q96` left to be "compensated" for, and thus you just divide by `q96` to cancel it out.

For `liquidity1()`, you only have the lot of `sqrtp` in the denominator, and to compensate for that you multiply by `q96` to cancel it out.

For the decimals, I think it's just for simplicity that 18 decimals are assumed for both tokens, and thus it's not the **real** USDC ERC20 definition. If you want to use 6 decimals (to be "true" to USDC), the differences in decimals will be reflected in the prices (which will account for any difference in decimals between the two tokens), and they'll all come out in the wash from the formula eventually.
